tests/trainer/test_tp.py

```diff
@@ -2,22 +2,333 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import contextlib
+from typing import Any, Optional, Union
 
+import numpy as np
 import pytest
 import torch
 from packaging import version
-from torch.utils.data import DataLoader
+from torch.distributed._tensor import Replicate
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.utils.data import DataLoader, Dataset
 
+from composer.callbacks import MemoryMonitor
+from composer.loggers import InMemoryLogger
 from composer.optim import DecoupledSGDW
 from composer.trainer.trainer import Trainer
-from composer.utils import dist
+from composer.utils import FSDPConfig, ParallelismConfig, TPConfig, dist, reproducibility
 from tests.common import (
     RandomClassificationDataset,
+    RandomClassificationDatasetReplicated,
     SimpleModel,
+    TPSimpleComposerMLP,
+    deep_compare,
     world_size,
 )
 
 
+def get_base_trainer(
+    parallelism_config: Optional[ParallelismConfig] = None,
+    size: int = 4,
+    batch_size: int = 1,
+    num_classes: int = 2,
+    num_features: int = 2,
+    seed: int = 44,
+    device: Union[torch.device, str] = 'cuda',
+    replication: Optional[int] = None,
+):
+    """Trainer for a simple model with any parallelism_config."""
+
+    reproducibility.seed_all(seed)
+    if isinstance(device, str):
+        device = torch.device(device)
+
+    dataset: Dataset = RandomClassificationDatasetReplicated(
+        shape=(num_features,),
+        num_classes=num_classes,
+        size=size,
+        device=device,
+        replication=replication,
+    )  # X=(num_features,), y=(,), i.e. scalar
+
+    dataloader = DataLoader(
+        dataset,
+        sampler=dist.get_sampler(dataset),
+        batch_size=batch_size,
+    )  # X=(batch_size, num_features), y=(batch_size,)
+
+    model = TPSimpleComposerMLP(num_features=num_features, device=device, num_classes=num_classes)
+
+    trainer = Trainer(
+        seed=seed,
+        device='gpu',
+        model=model,
+        max_duration='1ep',
+        train_dataloader=dataloader,
+        precision='fp32',
+        parallelism_config=parallelism_config,
+        callbacks=[MemoryMonitor()],
+        loggers=[InMemoryLogger()],
+        progress_bar=False,
+        log_to_console=False,
+    )
+
+    return trainer
+
+
+def get_trainer(
+    parallelism_strategy: str,
+    size: int = 4,
+    batch_size: int = 1,
+    num_classes: int = 2,
+    num_features: int = 2,
+    seed: int = 44,
+    device: Union[torch.device, str] = 'cuda',
+    replication: Optional[int] = None,
+) -> Trainer:
+
+    if parallelism_strategy == 'ddp':
+        trainer = get_base_trainer(
+            size=size,
+            batch_size=batch_size,
+            num_classes=num_classes,
+            num_features=num_features,
+            seed=seed,
+            device=device,
+            replication=replication,
+        )
```

Comment:
instead of calling `get_trainer` 3 times, it should be one call and the if statement just creates parallelism config. For DDP, it can be None

Reply:
Great idea! Done.